Change the .NET tool tutorial (#52952)
dotnet/docs: docs/core/tools/global-tools-how-to-create.md

ms.topic: tutorial
ms.date: 04/08/2026
---

# Tutorial: Create a .NET tool using the .NET CLI

**This article applies to:** ✔️ .NET 8 SDK and later versions

This tutorial teaches you how to create and package a .NET tool. The .NET CLI lets you create a console application as a tool, which others can install and run. .NET tools are NuGet packages that are installed from the .NET CLI. For more information about tools, see [.NET tools overview](global-tools.md).

The tool that you'll create is a console application that takes information about the current .NET environment and displays it, including the .NET version, operating system details, and key environment variable settings.

This is the first in a series of three tutorials. In this tutorial, you create and package a tool. In the next two tutorials you [use the tool as a global tool](global-tools-how-to-use.md) and [use the tool as a local tool](local-tools-how-to-use.md). The procedures for creating a tool are the same whether you use it as a global tool or as a local tool.

## Prerequisites

- [.NET SDK 10.0](https://dotnet.microsoft.com/download/dotnet/10.0) or a later version.

>   This tutorial uses .NET SDK 10.0. Global tools are available starting in .NET Core SDK 2.1, and local tools are available starting in .NET Core SDK 3.0.

Reviewer: Do we even need this paragraph anymore?

1. Navigate to the *repository* folder and enter the following command:

   ```dotnetcli
   dotnet new console -n dotnet-env
   ```

   The command creates a new folder named *dotnet-env* under the *repository* folder.

1. Navigate to the *dotnet-env* folder.

   ```console
   cd dotnet-env
   ```

## Add the code

1. Open the *Program.cs* file with your code editor.

1. Replace the contents with the following code:

   :::code language="csharp" source="./snippets/global-tools-how-to-create/csharp/Program.cs" id="fullprogram":::

   The program has two methods:

>    - The startup code checks displays the version of the command and then calls the `ShowInfo` method.

Reviewer: ```suggestion
   - The startup code checks the version of the command and then calls the `ShowInfo` method.
```

>    - `Main` checks whether any arguments were passed. With no arguments, it calls `ShowInfo()`. With any argument, it prints the tool version.

Author: ```suggestion
   - The startup code checks displays the version of the command and then calls the `ShowInfo` method.
```

   - `ShowInfo` displays three sections of information:
     - **Runtime** — the .NET version, framework description, and runtime identifier, using `Environment.Version` and `RuntimeInformation`.
     - **System** — OS description, architecture, machine name, and processor count.

>      - **Environment Variables** — six key .NET-related variables (`DOTNET_ROOT`, `DOTNET_HOST_PATH`, `DOTNET_CLI_HOME`, `DOTNET_NOLOGO`, `NUGET_PACKAGES`, and `DOTNET_ENVIRONMENT`), showing `(not set)` for any that aren't configured.

Reviewer: ```suggestion
     - **Environment variables** — six key .NET-related variables (`DOTNET_ROOT`, `DOTNET_HOST_PATH`, `DOTNET_CLI_HOME`, `DOTNET_NOLOGO`, `NUGET_PACKAGES`, and `DOTNET_ENVIRONMENT`), showing `(not set)` for any that aren't configured.
```

## Package the tool

To pack and distribute the application as a tool, modify the project file.

1. Open the *dotnet-env.csproj* file and add three new XML nodes to the end of the `<PropertyGroup>` node:

   :::code language="xml" source="./snippets/global-tools-how-to-create/csharp/dotnet-env.csproj" id="toolsettings":::

   `<ToolCommandName>` is an optional element that specifies the command that invokes the tool after installation. If this element isn't provided, the command name for the tool is the assembly name, which is typically the project file name without the *.csproj* extension.

   > [!NOTE]
   > Choose a unique value for `<ToolCommandName>`. Avoid using file extensions (like `.exe` or `.cmd`) because the tool is installed as an app host and the command shouldn't include an extension. This helps prevent conflicts with existing commands and ensures a smooth installation experience.

>    `<PackageOutputPath>` is an optional element that determines where the .NET CLI produces the NuGet package. The .NET CLI uses the NuGet package to install your tool.

Reviewer: ```suggestion
   `<PackageOutputPath>` is an optional element that determines where .NET produces the NuGet package. The .NET CLI uses the NuGet package to install your tool.
```
